### PR TITLE
Gradle api refactor

### PR DESCRIPTION
**Pull request by gmazzo**
Migrates to `Properties` API and `Lazy Configuration` task loading.

### CODE REVIEW

**Comment by gmazzo** [comment]
@ZacSweers  any feedback on this? Otherwise I'll release it soon as `3.0.0`

File: plugin/src/main/kotlin/com/github/gmazzo/gradle/plugins/BuildConfigPlugin.kt [hunk 1]
```diff
@@ -95,40 +101,53 @@ class BuildConfigPlugin : Plugin<Project> {
...
+                ?.let { "$it.${project.name}" }
+                ?: project.name
+        }
 
     companion object {
 
         const val DEFAULT_SOURCE_SET_NAME = "main"
 
+        @Suppress("DeprecatedCallableAddReplaceWith")
+        @Deprecated(message = "this should be Gradle's official `orElse`, but it's not available at Gradle 5")
```
Comment (ZacSweers) [inline plugin/src/main/kotlin/com/github/gmazzo/gradle/plugins/BuildConfigPlugin.kt:142]: Why support such old versions still if you're making a major/breaking update?

File: plugin/src/main/kotlin/com/github/gmazzo/gradle/plugins/generators/BuildConfigGeneratorSpec.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.github.gmazzo.gradle.plugins.generators
+
+import com.github.gmazzo.gradle.plugins.BuildConfigField
+import java.io.File
+
+data class BuildConfigGeneratorSpec(
```
Comment (ZacSweers) [inline plugin/src/main/kotlin/com/github/gmazzo/gradle/plugins/generators/BuildConfigGeneratorSpec.kt:6]: This should be a regular `abstract` class with gradle property APIs

**Review by ZacSweers — COMMENTED** [review]
Apologies for the delay in circling to this, it sort of disappeared from my radar. Left a few comments. Mostly good things here, I think you should be more ambitious with the changes though

**Pull request merged**